framework/midi/midievent.h

```diff
@@ -0,0 +1,956 @@
+        //! MIDI1.0 voice message
+        ChannelVoice10      = 0x2,
+        SystemExclusiveData = 0x3,
+        //! MIDI2.0 voice message
+        ChannelVoice20      = 0x4,
+        Data                = 0x5
+    };
+
+    enum Opcode {
+        RegisteredPerNoteController = 0b0000,
+        AssignablePerNoteController = 0b0001,
+        RegisteredController        = 0b0010,
+        AssignableController        = 0b0011,
+        RelativeRegisteredController= 0b0100,
+        RelativeAssignableController= 0b0101,
+        PerNotePitchBend            = 0b0110,
+        NoteOff                     = 0b1000,
+        NoteOn                      = 0b1001,
+        PolyPressure                = 0b1010,
+        ControlChange               = 0b1011,
+        ProgramChange               = 0b1100,
+        ChannelPressure             = 0b1101,
+        PitchBend                   = 0b1110,
+        PerNoteManagement           = 0b1111
+    };
+
+    enum AttributeType {
+        NoData = 0x00,
+        ManufacturerSpecific = 0x01,
+        ProfileSpecific = 0x02,
+        Pitch = 0x03
+    };
+
+    Event()
+        : m_data({ 0, 0, 0, 0 }) {}
+    Event(Opcode opcode, MessageType type = ChannelVoice20)
+    {
+        setMessageType(type);
+        setOpcode(opcode);
+    }
+
+    [[deprecated]] Event(channel_t ch, EventType type)
+    {
+        setType(type);
+        setChannel(ch);
+    }
+
+    [[deprecated]] Event(channel_t ch, EventType type, uint8_t a = 0, uint8_t b = 0)
+    {
+        m_data[0] = (a << 8) | b;
+        setType(type);
+        setChannel(ch);
+    }
+
+    //! 4.8.1 NOOP Utility message
+    static Event NOOP() { return Event(); }
+
+    //! from standard MIDI1.0 (specification 1996)
+    static Event fromMIDI10Package(uint32_t data)
+    {
+        Event e;
+        union {
+            unsigned char byte[4];
+            uint32_t full;
+        } u;
+        u.full = data;
+        if (
+            (u.byte[0] & 0x80)
+            || (u.byte[0] & 0x90)
+            || (u.byte[0] & 0xA0)
+            || (u.byte[0] & 0xB0)
+            || (u.byte[0] & 0xC0)
+            || (u.byte[0] & 0xD0)
+            || (u.byte[0] & 0xE0)
+            ) {
+            e.m_data[0] = (u.byte[0] << 16) | (u.byte[1] << 8) | u.byte[2];
+            e.setMessageType(ChannelVoice10);
+        }
+        return e;
+    }
+
+    uint32_t to_MIDI10Package() const
+    {
+        if (messageType() == ChannelVoice10) {
+            union {
+                unsigned char byte[4];
+                uint32_t uint32;
+            } p;
+            p.uint32 = m_data[0];
+            std::swap(p.byte[0], p.byte[2]);
+            p.byte[3] = 0;
+            return p.uint32;
+        }
+        return 0;
+    }
+
+    bool operator ==(const Event& other) const { return m_data == other.m_data; }
+    bool operator !=(const Event& other) const { return !operator==(other); }
+    operator bool() const {
+        return operator!=(NOOP());
+    }
```

Comment:
Is this operator really needed?
Would expect to see a method that `isValid()` (only more complex)

Reply:
NOOP - is a valid (and standard) MIDI message. In the same time, it is a NULL message, that may not be processed and can be skipped.
example of usage:
```
if (e) {
    m_eventReceived.send({ timing, e });
}
```
It won't litter the event stack.

Reply:
This is where I see problems

```
Event e1;
if (e1) {
  // not pass
}

Event e2;
e2.setMessageType(...);
if (e2) {
  // pass
  // but e2 not valid
}
```

Reply:
i.e. it is not obvious what the true means for the event

Reply:
Yes, I missed that case. I'll add a check for message's validation.

Reply:
fixed